### PR TITLE
Use new worker id on worker and client

### PR DESCRIPTION
**Pull request by dbw9580**
### What changes are proposed in this pull request?

Use new worker identity for worker registration and client-side cluster management.

### Why are the changes needed?

To allow workers to use generated IDs that are not assigned by the master.
To allow clients to build its view of the cluster using more efficient and reliable worker identity implementation.

### Does this PR introduce any user facing changes?

The worker identity is now persisted in etcd, if the worker uses etcd-based registration.

### CODE REVIEW

File: dora/core/common/src/main/java/alluxio/membership/WorkerServiceEntity.java
Comment (yyongycy) [inline dora/core/common/src/main/java/alluxio/membership/WorkerServiceEntity.java:68]: Wondering if a new constructor can be added to take the WorkerServiceEntity as an input parameter

File: dora/core/common/src/main/java/alluxio/membership/EtcdMembershipManager.java
Comment (yyongycy) [inline dora/core/common/src/main/java/alluxio/membership/EtcdMembershipManager.java:198]: https://github.com/Alluxio/alluxio/pull/18210/files#r1336587495 with constructor, this change can be avoided.

**Comment by JiamingMai** [comment]
image
There are still some failed tests that need to be fixed.

File: dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java
Comment (lucyge2022) [inline dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java:500]: oh is this an UnstableApi from guava?

File: dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java
Comment (lucyge2022) [inline dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java:383]: can u change name to GsonSerializer?

Comment (lucyge2022) [reply dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java:383]: is this serializer too much? why do we need such enum for serialize 2 fields?

File: common/transport/src/main/proto/grpc/block_master.proto
Comment (yyongycy) [inline common/transport/src/main/proto/grpc/block_master.proto:72]: If we treat this as the first version, we might consider to modify the first id directly.  
There is no backward compatible issue.
And is the first field still used for other part?

File: common/transport/src/main/proto/grpc/block_master.proto
Comment (yyongycy) [inline common/transport/src/main/proto/grpc/block_master.proto:71]: One more question, is the buildVersion used somewhere?

File: dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java
Comment (dbw9580) [reply dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java:383]: this is both a serializer and a deserializer.

> why do we need such enum for serialize 2 fields?

Most of the code is in the deserializer, which needs to handle cases like incorrect type of fields, fields missing, etc.

File: dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java
Comment (dbw9580) [reply dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java:602]: yes

File: common/transport/src/main/proto/grpc/block_master.proto
Comment (dbw9580) [reply common/transport/src/main/proto/grpc/block_master.proto:71]: I don't know

**Comment by dbw9580** [comment]
This PR is now split into two part: add serialization for worker identity which is #18224 , and the rest of code that make use of the identity, which is this PR.

commit [dcac2dc](https://github.com/Alluxio/alluxio/pull/18210/commits/dcac2dc0daa71b975fb97a8fd5fcb95d28147824) is the same as #18224. Please leave comments on this commit to that PR.

File: dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java
Comment (lucyge2022) [inline dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java:118]: so if this is used for serviceentityname which will eventaully becomes the path key on etcd, can u do "worker-%s" to avoid whitespaces

File: dora/core/common/src/main/java/alluxio/grpc/GrpcUtils.java
Comment (lucyge2022) [inline dora/core/common/src/main/java/alluxio/grpc/GrpcUtils.java:570]: so if workerinfo.id is marked deprecated, can we delete that setId() call?

File: dora/core/common/src/main/java/alluxio/grpc/GrpcUtils.java
Comment (lucyge2022) [inline dora/core/common/src/main/java/alluxio/grpc/GrpcUtils.java:571]: so im assuming this is used only for blockmaster calls to get worker list from master right? we don't need to support protobuf ser/de (toproto / from proto ) at all IMO.

File: dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java
Comment (lucyge2022) [inline dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java:59]: so there are 3 ser/de : jackson / gson / and java Serializable?
that's too much just for the purpose of ser/de, as WorkerIdentity is only used for membership module only, we dont need to consider anything else including master registration. if there's no usage, can u leave only the gson ser/de plz?

File: dora/core/server/worker/src/main/java/alluxio/worker/modules/LegacyWorkerIdentityProvider.java
Comment (lucyge2022) [inline dora/core/server/worker/src/main/java/alluxio/worker/modules/LegacyWorkerIdentityProvider.java:46]: where is this used? i didnt see it used in pageddoraworker registertomaster either? can u remove this if its not used? we dont want to write code only to serve the purpose of getting deprecated in the future

File: dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java
Comment (dbw9580) [reply dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java:118]: the `serviceentityname` is only used for display to humans, right? will it be consumed programmatically and parsed into anything else?

File: dora/core/common/src/main/java/alluxio/grpc/GrpcUtils.java
Comment (dbw9580) [reply dora/core/common/src/main/java/alluxio/grpc/GrpcUtils.java:571]: > don't need to support protobuf ser/de (toproto / from proto ) at all IMO.

Currently a lot of tests depend on protobuf ser/de. Unless we can ignore all these tests, I don't think this can be skipped.

File: dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java
Comment (dbw9580) [reply dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java:59]: Actually there are four: protobuf additionally.

Jackson is used to ser/de into/from `alluxio.wire.WorkerInfo` into JSON.
Gson is added recently to work with etcd.
java Serializable is implemented by `alluxio.wire.WorkerInfo`, which means anything embedded in it should also implement it.
Protobuf is used to ser/de into/from `alluxio.grpc.WorkerInfo`.

Comment (yyongycy) [reply dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java:59]: did you check in EE side? @dbw9580

File: dora/core/server/worker/src/main/java/alluxio/worker/modules/LegacyWorkerIdentityProvider.java
Comment (dbw9580) [reply dora/core/server/worker/src/main/java/alluxio/worker/modules/LegacyWorkerIdentityProvider.java:46]: Yes, this should be removed. It was a failed attempt to extract the worker registration process from `PagedDoraWorker` as a guice provider.

File: dora/core/common/src/main/java/alluxio/membership/StaticMembershipManager.java
Comment (JiamingMai) [inline dora/core/common/src/main/java/alluxio/membership/StaticMembershipManager.java:113]: Can worker identity configurable?

File: dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java
Comment (dbw9580) [reply dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java:59]: any other additional ser/de frameworks on that side?

Comment (yyongycy) [reply dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java:59]: one used for OPA de/serialize

File: common/transport/src/main/proto/grpc/block_master.proto
Comment (yyongycy) [reply common/transport/src/main/proto/grpc/block_master.proto:72]: @dbw9580  I am worrying this field is abandoned in the first day.

File: dora/core/common/src/main/java/alluxio/membership/WorkerServiceEntity.java
Comment (JiamingMai) [inline dora/core/common/src/main/java/alluxio/membership/WorkerServiceEntity.java:109]: Do we still need to compare the address?

File: dora/core/common/src/main/java/alluxio/membership/StaticMembershipManager.java
Comment (dbw9580) [reply dora/core/common/src/main/java/alluxio/membership/StaticMembershipManager.java:113]: as I noted in the todo comment above, it should be configurable by setting it in the static worker list file. Currently the file contains only worker net addresses, so the file should be changed to allow configuring both the identities and the net addresses at the same time.

Now I just generate the worker identities from the netaddresses. Do you think this is problematic?

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java
Comment (JiamingMai) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:307]: So it seems that master is still required though user can use ETCD as worker discovery service, righit?

**Review by JiamingMai — APPROVED** [review]
Mostly looks good. I left some comments.

File: dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java
Comment (dbw9580) [reply dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java:59]: is `WorkerInfo` or `WorkerIdentity` used in those code? At a first glance it does not seem to be related.

File: dora/core/common/src/main/java/alluxio/membership/WorkerServiceEntity.java
Comment (dbw9580) [reply dora/core/common/src/main/java/alluxio/membership/WorkerServiceEntity.java:109]: Good catch. Honestly I'm not very sure about the intent here. @lucyge2022 do you think we can remove `mAddress` from this class in favor of `mIdentity` ?

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java
Comment (dbw9580) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:307]: right

File: common/transport/src/main/proto/grpc/block_master.proto
Comment (dbw9580) [inline common/transport/src/main/proto/grpc/block_master.proto:62]: @jja725 do you know if any code related to PDDM or job service is still relying on this numeric worker ID? Can we have a list of items that need to be migrated to the new worker identity so that this field can be removed entirely?

File: dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java
Comment (dbw9580) [reply dora/core/common/src/main/java/alluxio/wire/WorkerIdentity.java:118]: changed to no whitespace in [c41d385](https://github.com/Alluxio/alluxio/pull/18210/commits/c41d385d1c756e2e164e699ba66166db553076ef)

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java
Comment (huanghua78) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:307]: Sorry, maybe I missed something.
Do you mean, `master` is still required to run in a Alluxio cluster even when we have ETCD service?

Comment (dbw9580) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:307]: For worker discovery purposes, either etcd or master is required, but not both.
For other features like PDDM, master is required.
If the user needs PDDM, then yes, etcd and master are required at the same time.

File: common/transport/src/main/proto/grpc/block_master.proto
Comment (jja725) [reply common/transport/src/main/proto/grpc/block_master.proto:62]: no, this can be removed since mostly we rely on workerNetAddress

File: dora/core/common/src/main/java/alluxio/membership/WorkerServiceEntity.java
Comment (lucyge2022) [reply dora/core/common/src/main/java/alluxio/membership/WorkerServiceEntity.java:109]: yeah u should definitely remove this, good catch

Comment (dbw9580) [reply dora/core/common/src/main/java/alluxio/membership/WorkerServiceEntity.java:109]: on a closer look, I think the worker net address should be preserved as the client needs it for creating connections. However, the netaddress should not be part of the equality check as it's only some supplementary data, not a defining factor of a `WorkerServiceEntity`.

Comment (dbw9580) [reply dora/core/common/src/main/java/alluxio/membership/WorkerServiceEntity.java]: excluded net address from the equality check in [1931f46](https://github.com/Alluxio/alluxio/pull/18210/commits/1931f467409efec00eb9dfd23bb3a94981a6dadb)

File: dora/core/client/fs/src/test/java/alluxio/client/file/dora/ConsistentHashPolicyTest.java
Comment (huanghua78) [inline dora/core/client/fs/src/test/java/alluxio/client/file/dora/ConsistentHashPolicyTest.java:90]: This was a bug?

**Comment by huanghua78** [comment]
I don't have further questions. Please address all the comments and get approvals before merge.

File: dora/core/client/fs/src/test/java/alluxio/client/file/dora/ConsistentHashPolicyTest.java
Comment (dbw9580) [reply dora/core/client/fs/src/test/java/alluxio/client/file/dora/ConsistentHashPolicyTest.java:90]: Not really. This PR changes the way worker identity is hashed, so the relative positions of the workers in the hash ring are affected. So this test should be updated accordingly.

**Review by lucyge2022 — APPROVED** [review]
LGTM

**Comment by dbw9580** [comment]
alluxio-bot, merge this please

**Pull request merged**
